File: backend/src/main/java/com/lsoft/aguapotable/controller/AlertaController.java
package com.lsoft.aguapotable.controller;

import com.lsoft.aguapotable.domain.entity.Alerta;
import com.lsoft.aguapotable.domain.entity.Usuario;
import com.lsoft.aguapotable.domain.enums.EstadoAlerta;
import com.lsoft.aguapotable.domain.repository.AlertaRepository;
import com.lsoft.aguapotable.domain.repository.UsuarioRepository;
import com.lsoft.aguapotable.exception.ResourceNotFoundException;
import com.lsoft.aguapotable.service.AlertaService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.servlet.http.HttpServletRequest;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import java.util.List;

@RestController
@RequestMapping("/api/v1/alertas")
@RequiredArgsConstructor
@Tag(name = "Alertas", description = "Gestión de alertas operativas")
@SecurityRequirement(name = "bearerAuth")
public class AlertaController {

    private final AlertaService alertaService;
    private final AlertaRepository alertaRepository;
    private final UsuarioRepository usuarioRepository;

    @GetMapping
    @Operation(summary = "Listar alertas de la comunidad (filtro por estado)")
    public ResponseEntity<List<Alerta>> listar(
            @RequestParam(required = false) EstadoAlerta estado,
            HttpServletRequest request) {
        Long comunidadId = (Long) request.getAttribute("comunidadId");
        if (estado != null) {
            return ResponseEntity.ok(alertaRepository.findByComunidadIdAndEstado(comunidadId, estado));
        }
        return ResponseEntity.ok(alertaService.getAlertasPendientes(comunidadId));
    }

    @PatchMapping("/{id}/resolver")
    @PreAuthorize("hasAnyRole('SUPER_ADMIN','ADMIN','OPERADOR')")
    @Operation(summary = "Marcar alerta como resuelta")
    public ResponseEntity<Alerta> resolver(@PathVariable Long id,
                                            HttpServletRequest request,
                                            Authentication authentication) {
        Long comunidadId = (Long) request.getAttribute("comunidadId");
        Alerta alerta = alertaRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Alerta", id));
        if (!alerta.getComunidad().getId().equals(comunidadId)) {
            throw new ResourceNotFoundException("Alerta", id);
        }

        Usuario usuario = usuarioRepository.findByEmail(authentication.getName()).orElse(null);
        alerta.setEstado(EstadoAlerta.RESUELTA);
        alerta.setFechaResuelta(LocalDateTime.now());
        alerta.setResueltaPor(usuario);

        return ResponseEntity.ok(alertaRepository.save(alerta));
    }

    @PatchMapping("/{id}/descartar")
    @PreAuthorize("hasAnyRole('SUPER_ADMIN','ADMIN')")
    @Operation(summary = "Descartar una alerta")
    public ResponseEntity<Alerta> descartar(@PathVariable Long id,
                                             HttpServletRequest request) {
        Long comunidadId = (Long) request.getAttribute("comunidadId");
        Alerta alerta = alertaRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Alerta", id));
        if (!alerta.getComunidad().getId().equals(comunidadId)) {
            throw new ResourceNotFoundException("Alerta", id);
        }
        alerta.setEstado(EstadoAlerta.DESCARTADA);
        return ResponseEntity.ok(alertaRepository.save(alerta));
    }
}

File: backend/src/alerta/alerta.controller.ts
import {
  Controller,
  Get,
  Patch,
  Param,
  Query,
  ParseIntPipe,
  UseGuards,
} from '@nestjs/common';
import { AuthGuard } from '@nestjs/passport';
import { ApiTags, ApiOperation, ApiBearerAuth } from '@nestjs/swagger';
import { AlertaService } from './alerta.service';
import { EstadoAlerta } from '../common/enums';
import { ComunidadId, CurrentUser } from '../common/decorators';
import { RolesGuard, Roles } from '../auth/guards/roles.guard';

@ApiTags('Alertas')
@ApiBearerAuth()
@Controller('api/v1/alertas')
@UseGuards(AuthGuard('jwt'), RolesGuard)
export class AlertaController {
  constructor(private readonly alertaService: AlertaService) {}

  @Get()
  @ApiOperation({ summary: 'Listar alertas de la comunidad (filtro por estado)' })
  async listar(
    @ComunidadId() comunidadId: number,
    @Query('estado') estado?: string,
  ) {
    if (estado) {
      return this.alertaService.listarPorEstado(
        comunidadId,
        estado as EstadoAlerta,
      );
    }
    return this.alertaService.getAlertasPendientes(comunidadId);
  }

  @Patch(':id/resolver')
  @Roles('SUPER_ADMIN', 'ADMIN', 'OPERADOR')
  @ApiOperation({ summary: 'Marcar alerta como resuelta' })
  async resolver(
    @Param('id', ParseIntPipe) id: number,
    @ComunidadId() comunidadId: number,
    @CurrentUser() user: any,
  ) {
    return this.alertaService.resolver(id, comunidadId, user.email);
  }

  @Patch(':id/descartar')
  @Roles('SUPER_ADMIN', 'ADMIN')
  @ApiOperation({ summary: 'Descartar una alerta' })
  async descartar(
    @Param('id', ParseIntPipe) id: number,
    @ComunidadId() comunidadId: number,
  ) {
    return this.alertaService.descartar(id, comunidadId);
  }
}
